keps/sig-api-machinery/0014-garbage-collection.md

```diff
@@ -0,0 +1,555 @@
+threshold on each of Kubernetes resources, and garbage collect resources after
+the threshold is hit. Note that only resources that are no longer in use can be
+cleaned up. 
+
+Points to consider when implementing the capacity-based garbage collection:
+
+* Fairness: for example, the capacity can be shared equally among namespaces.
+* Resource quota: capacity should be supplementary to resource quotas and can be
+  set according to resource quota.
+* Be generic and configurable.
+* Remove resources based on their last used time, instead of creation time.
+* Be coherent with other existing garbage collection mechanism.
+
+### Namespace-wide Garbage Collection Policy
+
+We can implement a garbage collection policy that applies to all resources in
+the same namespace by default for easier garbage collection policy management.
+
+### Move Event TTL out of etcd TTL
+
+We may consider moving Event TTL out of etcd TTL, and use the TTL mechanism in
+garbage collector for Events. 
+
+### Make Pod GC and Event GC Work Better with Garbage Collector
+
+Pod GC and Event GC don't respect OwnerReference-based garbage collector today.
+The biggest problem caused by Pod GC is that it breaks Job controller
+([#28486][]).
+
+We may move Pod GC and Event GC into garbage collector, or deprecate them and
+delete them entirely.
+
+### Move deletionGracePeriodSeconds from metadata to Pod spec
+
+`deletionGracePeriodSeconds` is currently in `.metadata` of all resources,
+although it only applies to Pods. This is confusing to the users.
+
+Because users cannot set or modify this field on a non-Pod resource (the field
+will be silently dropped), it will be compatible to move it outside of
+`.metadata` of resources other than Pod. This field can be part of a
+Pod-specific `.metadata`, or can even move this field to Pod spec in the future
+(in Pod v2 API).
+
+## Alternatives
+
+We have considered using owner references, instead of counting usage reference,
+to clean up resources after use. However, it has the following drawbacks:
+
+* Owner references of a resource may be excessively long. Resources like
+  ConfigMap can be referenced by Pods, ReplicaSets, and Deployments. If a
+  Deployment has 1k replicas, it will create 1k Pods and the ConfigMap needs to
+  add those 1k Pods to the ConfigMap’s owner references list. It will be easy
+  for the ConfigMap to hit Kubernetes resource object size limit. After the
+    limit is hit, the resource object cannot be updated anymore. 
+* Using owner references to implement this feature will make the feature harder
+  to be safely disabled after it’s enabled, because that requires all the added
+  owner references to be removed immediately before the garbage collector starts
+  reacting to it. It will cause resource objects to be cleaned up unexpectedly. 
+* Using owner references is conceptually more complex.
+
+We have also considered implementing a cluster wide cleanup policy. However we
```

Comment:
I believe cluster wide cleanup policy has some value as cluster admins can technically be users too.